### ISSUE TITLE
Serialization of 64bit values are truncated with .net sdk

### ISSUE THREAD

**Issue by kskem** [dapr/dapr#3837]
## In what area(s)?

/area runtime

> /area operator

> /area placement

> /area docs

> /area test-and-release

## What version of Dapr?

> 1.1.x
> 1.0.x
edge: output of `git describe --dirty`
starting Dapr Runtime -- version edge -- commit b1cad9a76e2ebbb5acee814f9897b47f43f9f4bb

## Expected Behavior

Sending 64bit values with pub/sub are expected to arrive unchanged.

## Actual Behavior

Receiving 64bit values are truncated in the trailing two decimal digits.
Best guess is that the serialized json has a max string-length for numbers.

## Steps to Reproduce the Problem

I have posted a reproducible setup, based on Docker, with instructions:
https://github.com/kskem/dapr-serialization-issue

## Release Note

RELEASE NOTE: **FIX** Bug in runtime.

**Comment by real-zony** [comment]
I have this problem too.

**Comment by hueifeng** [comment]
Any update?

**Comment by SayHelloGsx** [comment]
Yeah I got the same problem too. Could any one tell me how to fix it?

**Comment by yaron2** [comment]
I triaged this to be fixed in 1.6. 

cc @artursouza

**Comment by artursouza** [comment]
Is this a runtime or SDK issue?

**Comment by yaron2** [comment]
> Is this a runtime or SDK issue?

Runtime

**Comment by hueifeng** [comment]
Serialization is used in [runtime.go](https://github.com/dapr/dapr/blob/cb6b557bc016c21aaecdbf6d634e5e4145af2d62/pkg/runtime/runtime.go#L513), but when it encounters 64bit it automatically identifies Number as float64.

For example:

```
package main

import(
	jsoniter "github.com/json-iterator/go"
)

var name string

type Stu struct{
	Name string `json:"name"`
	Id int64
}

func main()  {
	Stu:=Stu{
		name:"张三",
		Id:637715205845980876
	}
	m1 :=make(map[string]interface{})
	var json= jsoniter.ConfigDefault
	var jsonStu,_=json.Marshal(stu)
	println(string(jsonStu))
	json.Unmarshal(jsonStu,&m1)

}
```
<img width="702" alt="739f6f2a275078a3e0fe9d341219d33" src="https://user-images.githubusercontent.com/34702552/146720430-875f3b77-3004-451c-b0f3-e8b349c58b44.png">

json.Unmarshal will use float64 to handle the problem, and you'll need to change the method to display the declaration.
```
var val map[string]interface{}
d := json.NewDecoder(bytes.NewBuffer(bdata))
d.UseNumber()
err = d.Decode(&val)
```

**Comment by yaron2** [comment]
Related, possible: https://github.com/json-iterator/go/issues/510

**Comment by hueifeng** [comment]
@yaron2 You can consider using the following code to solve.
```
package main

import (
	"strings"

	jsoniter "github.com/json-iterator/go"
)

type Stu struct {
	Name string `json:"name"`
	Id   int64
}

func main() {
	stu := Stu{
		Name: "张三",
		Id:   637715205845980876,
	}
	m1 := make(map[string]interface{})
	var json = jsoniter.ConfigDefault
	var jsonStu, _ = json.Marshal(stu)
	println(string(jsonStu))
	json.Unmarshal(jsonStu, &m1)

	decoder := jsoniter.NewDecoder(strings.NewReader(string(jsonStu)))
	decoder.UseNumber()
	d := make(map[string]interface{})
	decoder.Decode(&d)
	s2, _ := jsoniter.Marshal(d)
	println(string(s2))
}
```
Use the func (*Decoder) UseNumber method to deserialize the numeral type of json, instead of directly converting to float64, convert to json.Number, json.Number internal:
```
// A Number represents a JSON number literal.
type Number string

// String returns the literal text of the number.
func (n Number) String() string { return string(n) }

// Float64 returns the number as a float64.
func (n Number) Float64() (float64, error) {
    return strconv.ParseFloat(string(n), 64)
}

// Int64 returns the number as an int64.
func (n Number) Int64() (int64, error) {
    return strconv.ParseInt(string(n), 10, 64)
}
```

**Comment by yaron2** [comment]
Thanks @hueifeng.

@artursouza after looking at this, I recommend we move away from `jsoniter` to the Go stdlib `encoding/json` library. It looks like since Go 1.14 performance is no longer a critical aspect of this library: https://github.com/json-iterator/go/issues/455.

**Comment by artursouza** [comment]
> Thanks @hueifeng.
> 
> 
> 
> @artursouza after looking at this, I recommend we move away from `jsoniter` to the Go stdlib `encoding/json` library. It looks like since Go 1.14 performance is no longer a critical aspect of this library: https://github.com/json-iterator/go/issues/455.
> 
> 
> 
> Moving to `encoding/json` will allow us to optimize for JSON consistency and stability whilst removing a dependency.
> 
> 

LGTM

**Comment by yaron2** [comment]
@hueifeng would you like to contribute this change?

**Comment by hueifeng** [comment]
I can do it.

**Comment by artursouza** [comment]
E2E tests are not passing in the PR, moving to 1.7

**Comment by artursouza** [comment]
Moving to 1.7 because we are cutting RC for 1.6.

**Comment by shubham1172** [comment]
Dapr/dapr does not have a problem.  I tried putting telemetry around pub ([pkg/grpc/api.go](https://github.com/dapr/dapr/blob/9e970a0365181b5862aec06bf9902e4e72bc2278/pkg/grpc/api.go#L281)) and sub ([pkg/runtime/runtime.go](https://github.com/dapr/dapr/blob/9e970a0365181b5862aec06bf9902e4e72bc2278/pkg/runtime/runtime.go#L575)) code to check what data is being sent and received. 

Example data sent from application: 
```json
{"int32":1041701776,"uint32":1041701776,"int64":637831257272814404,"uint64":637831257272814404,"base64long":"RAtjLCYI2gg=","base64int":"kBsXPg=="}
```

On Dapr side, while sending the data, the raw bytes from Marshaling the struct gets its `int64` and `uint64` data converted to float, and here it loses precision. Sample data after marshaling on publisher's end:
```go
map[data:map[base64int:kBsXPg== base64long:RAtjLCYI2gg= int32:1.041701776e+09 int64:6.378312572728145e+17 uint32:1.041701776e+09 uint64:6.378312572728145e+17]
```

On Dapr side, while receiving the data, the raw bytes itself have incorrect data (as in [shubham1172/dapr-serialization-issue](https://github.com/shubham1172/dapr-serialization-issue/)'s readme). Thus Unmarshaling with `UseNumber` here won't help. Sample data after unmarshalling:
```go
map[data:map[base64int:kBsXPg== base64long:RAtjLCYI2gg= int32:1041701776 int64:637831257272814500 uint32:1041701776 uint64:637831257272814500]

The root cause is the `json.Unmarshal` in https://github.com/dapr/components-contrib/blob/master/pubsub/envelope.go
On adding the `UseNumber` way, it works as expected without data loss. I will create a PR and attach to this issue soon.

Should we create a separate item to move away from jsoniter to encoding/json?

**Comment by shubham1172** [comment]
I did some more research, and this will **not** be a breaking change. 

The only change that this PR introduces is in the unmarshalling of the CloudEvent data from the PublishEventRequest in PublishEvent ([api.go](https://github.com/dapr/dapr/blob/2062c3ed344759e213d0f3b4a05df35283fcbcd4/pkg/grpc/api.go#L235))

Earlier, this unmarshalling resulted in data loss while converting 64-bit values to 53-bits float (as per [JSON](https://datatracker.ietf.org/doc/html/rfc7159)). Now, it uses `json.Number` representation, an alias of string, to unmarshall the same without any data loss.

After unmarshalling the CloudEvent's data, it's packed into the PublishRequest as `[]byte` (both HTTP and GRPC). This is the request that gets published on the actual pubsub component.

The same thing is being read as `[]byte` by our subscribers, in beginPubSub ([runtime.go](https://github.com/dapr/dapr/blob/2062c3ed344759e213d0f3b4a05df35283fcbcd4/pkg/runtime/runtime.go#L560))

Since the public API contracts are not changed, this won't affect anything external.

**Comment by shubham1172** [comment]
@mukundansundar can you please un-tag the PR as `breaking-change`?

**Comment by vinceyuan** [comment]
I still see this issue in Go `1.18.5`. `295219529439528020` is recognized as `295219529439528000`.
